core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/OutputExpression.scala

```diff
@@ -1,31 +1,40 @@
 package org.broadinstitute.dsde.rawls.expressions
 
 import org.broadinstitute.dsde.rawls.{RawlsException, StringValidationUtils}
-import org.broadinstitute.dsde.rawls.model.{AttributeName, ErrorReportSource}
+import org.broadinstitute.dsde.rawls.model.{Attribute, AttributeName, ErrorReportSource}
 
-sealed trait OutputExpressionTarget {
-  val prefix: String
-  def remainder(s: String): String = s.stripPrefix(prefix)
-}
-case object ThisEntityTarget extends OutputExpressionTarget { override val prefix = "this." }
-case object WorkspaceTarget extends OutputExpressionTarget { override val prefix = "workspace." }
+import scala.util.{Failure, Try}
+
+sealed trait OutputExpressionTarget { val root: String }
+case object ThisEntityTarget extends OutputExpressionTarget { override val root = "this." }
+case object WorkspaceTarget extends OutputExpressionTarget { override val root = "workspace." }
 
-case class OutputExpression(target: OutputExpressionTarget, attributeName: AttributeName)
+sealed trait OutputExpression
+case object UnboundOutputExpression extends OutputExpression
+case class BoundOutputExpression(target: OutputExpressionTarget, attributeName: AttributeName, attribute: Attribute) extends OutputExpression
 
-object OutputExpression extends StringValidationUtils {
+object BoundOutputExpression extends StringValidationUtils {
   override implicit val errorReportSource = ErrorReportSource("rawls")
 
-  def apply(s: String): OutputExpression = {
-    val target = if(s.startsWith(ThisEntityTarget.prefix))
-      ThisEntityTarget
-    else if(s.startsWith(WorkspaceTarget.prefix))
-      WorkspaceTarget
-    else
-      throw new RawlsException(s"Invalid output expression: $s")
+  def tryParse(target: OutputExpressionTarget, expr: String, attribute: Attribute): Try[BoundOutputExpression] = {
+    if (expr.startsWith(target.root)) Try {
+      val attributeName = AttributeName.fromDelimitedName(expr.stripPrefix(target.root))
+      validateUserDefinedString(attributeName.name)
+
+      BoundOutputExpression(target, attributeName, attribute)
+    }
+    else Failure(new RawlsException(s"Invalid output expression: $expr"))
+  }
 
-    val attributeName = AttributeName.fromDelimitedName(target.remainder(s))
-    validateUserDefinedString(attributeName.name)
+  def apply(expr: String, attribute: Attribute): BoundOutputExpression = {
+    val parseResult = tryParse(ThisEntityTarget, expr, attribute) recoverWith { case _ => tryParse(WorkspaceTarget, expr, attribute) }
+    parseResult.get
```

Comment:
`BoundOutputExpression("blarg.blorg", AttributeNumber(2))` will throw an exception, no? which you catch in SMA line 350 by wrapping it in a `Try {}` block, but that's a burden on the caller. perhaps make this more explicit by replacing your `apply` functions with an explict `OutputExpression.from(): Try[OutputExpression]`. it'd be ok to call `get` on that in SMA but that way other people don't get surprised by exceptions